airbyte-cdk/python/connector_builder/message_grouper.py

```diff
@@ -101,33 +103,50 @@ def _get_message_groups(
                 current_page_response = self._create_response_from_log_message(message.log)
             elif message.type == Type.LOG:
+                if message.log.level == Level.ERROR:
+                    had_error = True
                 yield message.log
             elif message.type == Type.RECORD:
                 current_page_records.append(message.record.data)
                 records_count += 1
                 schema_inferrer.accumulate(message.record)
         else:
-            self._close_page(current_page_request, current_page_response, current_slice_pages, current_page_records)
+            self._close_page(current_page_request, current_page_response, current_slice_pages, current_page_records, validate_page_complete=not had_error)
             yield StreamReadSlices(pages=current_slice_pages)
 
     @staticmethod
-    def _need_to_close_page(at_least_one_page_in_group, message):
+    def _need_to_close_page(at_least_one_page_in_group, message) -> bool:
         return (
                 at_least_one_page_in_group
                 and message.type == Type.LOG
                 and (message.log.message.startswith("request:") or message.log.message.startswith("slice:"))
         )
 
     @staticmethod
-    def _close_page(current_page_request, current_page_response, current_slice_pages, current_page_records):
-        if not current_page_request or not current_page_response:
+    def _close_page(current_page_request, current_page_response, current_slice_pages, current_page_records, validate_page_complete: bool):
+        """
+        Close a page when parsing message groups
+        @param validate_page_complete: in some cases, we expect the CDK to not return a response. As of today, this will only happen before
+        an uncaught exception and therefore, the assumption is that `validate_page_complete=True` only on the last page that is being closed
+        """
+        if validate_page_complete and (not current_page_request or not current_page_response):
             raise ValueError("Every message grouping should have at least one request and response")
 
         current_slice_pages.append(
             StreamReadPages(request=current_page_request, response=current_page_response, records=deepcopy(current_page_records))
         )
         current_page_records.clear()
 
+    def _read_stream(self, source, config, configured_catalog) -> Iterator[AirbyteMessage]:
+        # the generator can raise an exception
+        # iterate over the generated messages. if next raise an exception, catch it and yield it as an AirbyteLogMessage
+        try:
+            yield from source.read(logger=self.logger, config=config, catalog=configured_catalog, state={})
+        except Exception as e:
+            error_message = f"{e.args[0] if len(e.args) > 0 else str(e)} - {ErrorFormatter.get_stacktrace_as_string(e)}"
+            yield AirbyteMessage(type=MessageType.LOG, log=AirbyteLogMessage(level=Level.ERROR, message=error_message))
```

Comment:
Out of curiosity, why are we returning a `MessageType.LOG` here instead of `MessageType.TRACE`?

Reply:
We're returning it as a log message so it can be [appended to the list of log messages](https://github.com/airbytehq/airbyte/pull/24173/files#diff-01d26032991ab687a8a2142afa745bbb0b92e4cb190835a19c8f3b151a10c699R52).

You bring up a good point though. It might make sense to return a trace message and also add these to the StreamRead's logs

Reply:
Yeah the advantage of that is that it might be nice for the Connector Builder server to parse it as a trace message rather than having to look for errors in a log message type, when deciding what kind of response to send to the client.